omnibus/files/private-chef-ctl-commands/cleanup_bifrost.rb

```diff
@@ -0,0 +1,213 @@
+  install_bifrost_tracking_table(bifrost_db)
+  # NOTE(ssd) 2017-07-24:
+  # For the most part, the tracking table manages in-flight requests for us.
+  # However, it is still possible that the following sequence happens:
+  #    - new authz_id in bifrost database
+  #    - tracking table inserted
+  #    - known_actor_list fetched
+  #    - new record in oc-chef database
+  #
+  # In the this case, the authz_id won't be in our list of known
+  # clients, won't be in our tracking table, but will show up in our
+  # search for orphaned authz_ids.
+  #
+  # Thus, this sleep is a 1-time wait to avoid that scenario.
+  #
+  puts "One-time sleep to account for in-flight requests not captured by tracking table"
+  sleep 25
+  begin
+    estimated_deletion_count = print_and_return_estimate(known_actor_list, bifrost_db)
+
+    if estimated_deletion_count <= 0
+      puts "Estimated deletion count 0. Aborting"
+      exit(0)
+    end
+
+    run_bifrost_scan(known_actor_list, batch_size, bifrost_db)
+  ensure
+    remove_bifrost_tracking_table(bifrost_db)
+  end
+end
+
+def known_actor_list
+  @known_actor_list ||= timed "Fetching initial opscode_chef clients and user list" do
+    clients_and_users = erchef_db.exec("SELECT authz_id FROM clients UNION select authz_id FROM users")
+    clients_and_users.map do |real_actor|
+      real_actor['authz_id']
+    end
+  end
+end
+
+def safety_check(db)
+  res = db.exec("SELECT * FROM pg_tables
+                          WHERE tablename='cleanup_tracking_auth_actors'")
+  if res.ntuples > 0
+    puts "ERROR: cleanup_tracking_auth_actors already exists.  cleanup-bifrost may be running."
+    puts "ERROR: If you are sure cleanup-bifrost is not running, you can clean up the tracking tables with: chef-backend-ctl cleanup-bifrost --force-cleanup"
+  end
+end
+
+def install_bifrost_tracking_table(db)
+  puts "Installing tracking tables into bifrost database"
+  db.exec(CREATE_SQL)
+end
+
+def remove_bifrost_tracking_table(db)
+  puts "Removing tracking tables into bifrost database"
+  db.exec(DELETE_SQL)
+end
+
+def print_and_return_estimate(known_actor_list, db)
+  tcount = timed "Fetchings all actor counts" do
+    db.exec_params("SELECT count(*) FROM auth_actor").first['count'].to_i
+  end
+
+  estimated_del_count = [0, tcount - known_actor_list.length].max
+  puts "\n----------------------------------------"
+  puts " Total chef users+clients: #{known_actor_list.length}"
+  puts "Total bifrost auth_actors: #{tcount}"
+  puts "Deletion Candidates (est): #{estimated_del_count}"
+  puts "----------------------------------------\n"
+
+  estimated_del_count
+end
+
+def run_bifrost_scan(known_actor_list, batch_size, db)
+  total_deleted = 0
+  known_actors_str = to_pg_list(known_actor_list)
+  loop do
+    puts "Processing batch of #{batch_size}"
+
+    candidates = timed "    Retrieving batch from auth_actors" do
+      db.exec("SELECT authz_id FROM auth_actor WHERE authz_id NOT IN (#{known_actors_str}) LIMIT $1",
```

Comment:
Just curious -- there's no way to template list args in SQL, I suppose? (We're using `$1` for `batch_size`, but string interpolation for `known_actors_str`, here, and `to_pg_list(candidates)` below)

(I'm not worried about SQLi there, really, just curious)